File: conformance/tests/loop-if-merged-locals/LoopIfMergedLocals.runar.java
package runar.conformance.loopifmergedlocals;

import runar.lang.StatefulSmartContract;
import runar.lang.annotations.Public;
import runar.lang.types.Bigint;

/**
 * LoopIfMergedLocals -- branch-merged locals whose merged values are DEAD in
 * the enclosing scope, which is what an {@code if} inside a loop body always
 * makes them. Companion to {@code merge-locals-shapes} (merged locals LIVE
 * after the {@code if}) and {@code bounded-loop} (a loop with no branch in it)
 * -- neither covers their intersection, which is where the merge block's
 * premise fails.
 */
class LoopIfMergedLocals extends StatefulSmartContract {

    Bigint a;
    Bigint b;
    Bigint c;

    LoopIfMergedLocals(Bigint a, Bigint b, Bigint c) {
        super(a, b, c);
        this.a = a;
        this.b = b;
        this.c = c;
    }

    @Public
    void guarded(Bigint x, Bigint limit) {
        Bigint na = Bigint.ZERO;
        Bigint nb = Bigint.ZERO;
        for (Bigint i = Bigint.ZERO; i.lt(Bigint.of(2)); i = i.plus(Bigint.ONE)) {
            if (i.lt(limit)) {
                na = na.plus(x);
                nb = nb.plus(na);
            }
        }
        this.addOutput(1000L, na, nb, this.c);
    }

    @Public
    void afterIf(Bigint x, Bigint limit) {
        Bigint na = Bigint.ZERO;
        Bigint nb = Bigint.ZERO;
        for (Bigint i = Bigint.ZERO; i.lt(Bigint.of(2)); i = i.plus(Bigint.ONE)) {
            if (i.lt(limit)) {
                na = na.plus(x);
            }
            nb = nb.plus(na);
        }
        this.addOutput(1000L, na, nb, this.c);
    }
}

File: conformance/tests/loop-if-merged-locals/LoopIfMergedLocals.runar.go
//go:build ignore

package x

import runar "github.com/icellan/runar/packages/runar-go"

// LoopIfMergedLocals -- branch-merged locals whose merged values are DEAD in
// the enclosing scope, which is what an `if` inside a loop body always makes
// them. Companion to `merge-locals-shapes` (merged locals LIVE after the `if`)
// and `bounded-loop` (a loop with no branch in it) -- neither covers their
// intersection, which is where the merge block's premise fails.
type LoopIfMergedLocals struct {
	runar.StatefulSmartContract

	A runar.Bigint
	B runar.Bigint
	C runar.Bigint
}

func (c *LoopIfMergedLocals) Guarded(x runar.Bigint, limit runar.Bigint) {
	na := runar.Bigint(0)
	nb := runar.Bigint(0)
	for i := runar.Bigint(0); i < 2; i++ {
		if i < limit {
			na = na + x
			nb = nb + na
		}
	}
	c.AddOutput(1000, na, nb, c.C)
}

func (c *LoopIfMergedLocals) AfterIf(x runar.Bigint, limit runar.Bigint) {
	na := runar.Bigint(0)
	nb := runar.Bigint(0)
	for i := runar.Bigint(0); i < 2; i++ {
		if i < limit {
			na = na + x
		}
		nb = nb + na
	}
	c.AddOutput(1000, na, nb, c.C)
}
